### PR TITLE
Support arbitrary number of feature dimensions in latent to discrete transformation

### PR DESCRIPTION
**Pull request by VolodyaCO**
The encoder might not encode a datapoint x to a 1D spin string. In the case of autoencoders where vector quantisation layers are added, encoders might encode a datapoint x to a HxWxC array. In such a case, the default latent to discrete transformation would be wrong, as it expected only a single feature dimension, whereas in this case there are three such feature dimensions.

This PR fixes that by automatically finding the number of feature dimensions in the encoded data representation, and performs the gumbel softmax transformation accordingly, keeping the same number of feature dimensions.

### CODE REVIEW

File: dwave/plugins/torch/models/discrete_variational_autoencoder.py [hunk 1]
```diff
@@ -80,28 +80,26 @@ def __init__(
         self._decoder = decoder
         if latent_to_discrete is None:
 
-            def latent_to_discrete(
-                logits: torch.Tensor, n_samples: int
-            ) -> torch.Tensor:
-                # Logits is of shape (batch_size, n_discrete), we assume these logits
+            def latent_to_discrete(logits: torch.Tensor, n_samples: int) -> torch.Tensor:
+                # Logits is of shape (batch_size, f_1, f_2, ...), we assume these logits
```
Comment (thisac) [inline dwave/plugins/torch/models/discrete_variational_autoencoder.py:84]: Why `f_1`, etc. instead of `l1`, etc. as in the docstring above?

**Review by thisac — COMMENTED** [review]
Thanks @VolodyaCO. Don't forget to add a an entry to the release-log.

**Comment by VolodyaCO** [comment]
@thisac I don't understand very well the reno process for release notes. Is every PR intended to create a new release?

**Review by anahitamansouri — APPROVED** [review]
Thanks very much Vlad for the nice PR. I've added a few minor comments.

File: tests/test_dvae_winci2020.py [hunk 2]
```diff
@@ -154,3 +183,5 @@ def test_forward(self, n_samples):
 
 if __name__ == "__main__":
     unittest.main()
+    unittest.main()
+    unittest.main()
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply tests/test_dvae_winci2020.py:187]: Ah, yes, it is! thanks for noticing.

File: tests/test_dvae_winci2020.py [hunk 3]
```diff
@@ -48,16 +47,45 @@ def setUp(self):
...
+                return output
+
+        class Decoder(torch.nn.Module):
+            def __init__(self, latent_features: int, input_features: int):
+                super().__init__()
+                self.linear = torch.nn.Linear(latent_features, input_features)
+
             def forward(self, x: torch.Tensor) -> torch.Tensor:
-                return x * 20 - 10
+                # x is of shape (batch_size, replica_size, l1, l2, ...)
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply tests/test_dvae_winci2020.py]: Changed this to `n_samples`.

**Review by kevinchern — APPROVED** [review]
LGTM with some very minor feedback

File: tests/test_dvae_winci2020.py [hunk 4]
```diff
@@ -114,38 +144,43 @@ def test_train(self):
...
+        [
+            (1, torch.tensor([[[1.0, 1.0]], [[1.0, -1.0]], [[-1.0, -1.0]], [[-1.0, 1.0]]])),
+            (
+                5,
+                torch.tensor(
+                    [[[1.0, 1.0]] * 5, [[1.0, -1.0]] * 5, [[-1.0, -1.0]] * 5, [[-1.0, 1.0]] * 5]
+                ),
+            ),
+        ]
+    )
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply tests/test_dvae_winci2020.py]: This is the same. I'll leave it like it is, or you can commit this suggestion later.

**Review by thisac — APPROVED** [review]
LGTM
